site_scons/buildutils.py

```diff
@@ -1,78 +1,160 @@
+    10 and the highest, CRITICAL, set to 50. This filter causes a log handler to
+    reject messages that are above or below numerical cutoffs. Example::
+
+    >>> # Handles log levels from debug up to, but not including, error
+    >>> handler.addFilter(LevelFilter(logging.DEBUG, logging.ERROR))
+    >>> # Handles log levels from warning up to and including critical
+    >>> handler.addFilter(LevelFilter(logging.WARNING, logging.CRITICAL+1))
+    """
+
+    def __init__(self, low: int, high: int) -> None:
+        self._low = low
+        self._high = high
+        super().__init__()
+
+    def filter(self, record: logging.LogRecord) -> bool:
+        if self._low <= record.levelno < self._high:
+            return True
+        return False
+
+
+logger = logging.getLogger("cantera")
+logger.setLevel(logging.INFO)
+f = OutputFormatter()
+stdout_handler = logging.StreamHandler(sys.stdout)
+stdout_handler.setFormatter(f)
+stdout_handler.addFilter(LevelFilter(logging.DEBUG, logging.ERROR))
+logger.addHandler(stdout_handler)
+
+stderr_handler = logging.StreamHandler(sys.stderr)
+stderr_handler.setFormatter(f)
+stderr_handler.addFilter(LevelFilter(logging.ERROR, logging.CRITICAL + 1))
+logger.addHandler(stderr_handler)
+
+
+if sys.version_info[:2] < (3, 6):
+    logger.error(
+            "Cantera must be built using Python 3.6 or higher. If you are using "
+            "a system install of SCons, you can invoke SCons by writing\n"
+            "    python3 `which scons`\n"
+            "followed by any desired options."
+        )
+    sys.exit(1)
+
+
+class TestState(enum.IntEnum):
+    PASS = 0
+    FAIL = 1
+
+
+class DefineDict:
     """
     A dictionary-like object which generates appropriate preprocessor
     define statements from its dict of variable / value
     pairs. Variables whose value is None or that are not in the dict
     are left undefined.
     """
-    def __init__(self, data):
+    def __init__(self, data: dict) -> None:
         self.data = data
         self.undefined = set()
 
-    def __getitem__(self, key):
-        if key not in self.data:
+    def __getitem__(self, key: str) -> str:
+        if key not in self.data or self.data[key] is None:
             self.undefined.add(key)
-            return '/* #undef %s */' % key
-        elif self.data[key] is None:
-            return '/* #undef %s */' % key
+            return f"/* #undef {key!s} */"
         else:
-            return '#define %s %s' % (key, self.data[key])
+            return f"#define {key!s} {self.data[key]!s}"
 
 
-class ConfigBuilder(object):
+class ConfigBuilder:
     """
     Used along with DefineDict to generate a customized config.h file
     from a config.h.in file using the variables given in 'defines'.
     """
-    def __init__(self, defines):
+    def __init__(self, defines: dict) -> None:
         self.defines = DefineDict(defines)
 
-    def __call__(self, source, target, env):
+    def __call__(self, target, source, env):
```

Comment:
Good point... I'm interested in your opinion about this. The reason I left them out of the functions that I did here is that they seemed more like noise and not so helpful for future use. It would require importing several libraries under the type hinting check, and I didn't think it would add enough value to include the types in the signature. That may change as I go through the other files, though. What do you think?

Reply:
I *think* you can keep this relatively simple by using type aliases, i.e. `TypeVar` (see [here](https://www.python.org/dev/peps/pep-0484/#id17))
```
T = TypeVar('T', bound=...)
```
I.e. you just locally define an alias that documents what this? I had to dig this up from something one of my students introduced at some point.

Reply:
I don't think it's necessary to use `TypeVar` here, a simple alias would be enough. I'll add types in a single commit, and let's see if it doesn't make the code illegible. If it looks terrible, I can just drop that commit.

Reply:
Alright, it didn't come out too bad. I didn't add types for unused arguments, but everything else should have one now.

Reply:
It looks great, within the context that the aesthetics of PEP 484 may not be universally liked (I myself don't mind the annotations; on the contrary, I like the fact that it really helps when looking at an interface that you may not be familiar with).

Regarding unused arguments (e.g. 'target' and 'source' in `regression_test_message`), it may still help to know why they remain there (and what types they would be). You mentioned this in `print_report`, but it still is unclear what those required arguments represent.